New OQD end-to-end pipeline is added to Catalyst

rniczh opened the pull request:
**Context:**

**Description of the Change:**

This pipeline enables Catalyst to compile circuit for ARTIQ-based quantum device. When `device_db` configuration is provided, the compilation follows the ARTIQ route:
- `ions-decomposition`: Decompose quantum operations for trapped ion systems
- `gates-to-pulses`: Convert gate operations to pulse sequences
- `convert-ion-to-rtio`: Convert ion operations to RTIO dialect
- `convert-rtio-event-to-artiq`: Lower RTIO event to ARTIQ's primitives
- `llvm-dialect-lowering-stage`: Lower to LLVM IR
- `emit-artiq-runtime`: Generate ARTIQ runtime entry point as wrapper for ARITQ device to execute

The final stage compiles LLVM IR to an ELF binary targeting the ARTIQ device (ARM Cortex-A9). Due to current limitations where Catalyst's internal LLVM build does not include the corresponding ARM backend, the compilation will use `llc` to compile `.ll` to object file, and use `ld.lld` to compile to `.elf`

Note: This PR only covers LLVM IR generation; the `compile_to_artiq` function itself is not included.


***Example:***
Replace the artiq configs to your own env setting:
```python
import os

import numpy as np
import pennylane as qml

from catalyst import QJIT, CompileOptions
from catalyst.third_party.oqd import OQDDevice, OQDDevicePipeline

OQD_PIPELINES = OQDDevicePipeline(
    os.path.join("calibration_data", "device.toml"),
    os.path.join("calibration_data", "qubit.toml"),
    os.path.join("calibration_data", "gate.toml"),
    os.path.join("device_db", "device_db.json"),
)


def test_rx_gate():
    """Test RX gate with ARTIQ linking done in user code."""
    oqd_dev = OQDDevice(
        backend="default",
        shots=4,
        wires=1
    )
    qml.capture.enable()

    # Compile to LLVM IR only
    @qml.qnode(oqd_dev)
    def circuit():
        x = np.pi / 2
        qml.RX(x, wires=0)
        return qml.counts(wires=0)
        
    compiled_circuit = QJIT(circuit, CompileOptions(link=False, pipelines=OQD_PIPELINES))

    # Compile to ARTIQ ELF (using `compile_to_artiq` function from third-party repo to compile to it)
    artiq_config = {
        "kernel_ld": "/path/to/kernel.ld",
        "llc_path": "/path/to/llc",
        "lld_path": "/path/to/ld.lld",
    }

    output_elf_path = compile_to_artiq(compiled_circuit, artiq_config)
    print(f"ARTIQ ELF file generated: {output_elf_path}")

test_rx_gate()
```

The result will store to `circuit.elf`

```
[ARTIQ] Generated ELF: circuit.elf
```

And you can run it on artiq device:

```shell
artiq_run  --device-db device_db.py circuit.elf
```

**Benefits:**

**Possible Drawbacks:**
~It causes the OQD specific compile stuffs steps into the original catalyst compiler driver's pipepline~

**Related GitHub Issues:**
[sc-100853]

rniczh committed:
New OQD end-to-end pipeline is added to Catalyst

rniczh committed:
formatting

rniczh committed:
update comment

rniczh committed:
inconsistent return statements

rniczh committed:
update comment

rniczh committed:
Merge branch 'main' of github.com:PennyLaneAI/catalyst into rniczh/add-new-oqd-pipeline

rniczh committed:
fix

rniczh committed:
add missing lib

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
Update

rniczh committed:
pylint

rniczh committed:
add tests for EmitArtiqRuntime

rniczh committed:
move the artiq specific logic out catalyst

rniczh committed:
udpate

rniczh committed:
revert

rniczh committed:
update

rniczh committed:
revertT

rniczh committed:
codecov

rniczh committed:
pylint

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
date

rniczh committed:
revert redundant change

paul0403 commented on mlir/lib/RTIO/Transforms/RTIOEmitARTIQRuntime.cpp:
```diff
@@ -0,0 +1,164 @@
+
+  private:
+    /// Emit ARTIQ runtime wrapper for LLVM dialect kernel function
+    LogicalResult emitARTIQRuntimeForLLVMFunc(ModuleOp module, OpBuilder &builder,
```
`module` is a keyword in cpp, let's avoid it as a variable name maybe

rniczh replied:
Done https://github.com/PennyLaneAI/catalyst/pull/2299/commits/9f95ddd6d294ebc00d86762c2a3b70b75df9153f

paul0403 reviewed (commented):
Thanks Hongsheng, looks really nice! I think this is the minimally intrusive way we can do it without altering too much of the base jit flow!

I left some comments but all are minor. Another thing is can you add your PR description's example as an end-to-end pytest in the frontend oqd tests? Obviously no need for execution, just check that the compiled ELF exist (maybe check a little bit of its content as well)? You may need to do things like `subprocess.run("which llc", shell=True)` in the test, but I think that's possible.

rniczh committed:
Update mlir/lib/RTIO/Transforms/RTIOEmitARTIQRuntime.cpp

Co-authored-by: Paul <79805239+paul0403@users.noreply.github.com>

mehrdad2m commented on frontend/catalyst/jit.py:
```diff
@@ -546,6 +546,7 @@ def __init__(self, fn, compile_options):
         self.mlir_module = None
         self.out_type = None
         self.overwrite_ir = None
+        # Use cwd for workspace if keep_intermediate is set
```
```suggestion
```
This comment seems redundant since the code is very verbose itself

mehrdad2m reviewed (commented):
Thanks @rniczh, I did a first round of review with some minor comments. However, I would like to play around with a bit more and maybe more review on Monday.

rniczh committed:
Update jit.py

Co-authored-by: Mehrdad Malek <39844030+mehrdad2m@users.noreply.github.com>

rniczh committed:
Update jit.py

Co-authored-by: Mehrdad Malek <39844030+mehrdad2m@users.noreply.github.com>

rniczh committed:
Update oqd_compile.py

Co-authored-by: Mehrdad Malek <39844030+mehrdad2m@users.noreply.github.com>

rniczh committed:
remove verbose

rniczh committed:
rename var from module to moduleOp:

rniczh committed:
revert copyright date

rniczh commented on frontend/catalyst/third_party/oqd/__init__.py:
```diff
@@ -1,4 +1,4 @@
-# Copyright 2024 Xanadu Quantum Technologies Inc.
+# Copyright 2024-2026 Xanadu Quantum Technologies Inc.
```
Done https://github.com/PennyLaneAI/catalyst/pull/2299/commits/9f95ddd6d294ebc00d86762c2a3b70b75df9153f

rniczh committed:
refactor subprocess call

rniczh committed:
Update Makefile and Add test

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
fix codecov

rniczh committed:
Add lld only when ENABLE_OQD=ON

rniczh committed:
Revert "Add lld only when ENABLE_OQD=ON"

This reverts commit 5b425528ead901dfa5bdf3f147b099b9086eb39d.

rniczh committed:
Revert "fix codecov"

This reverts commit 4798a9e6d250bc3cf3e42a1aff3320c979d2b031.

rniczh committed:
Revert "Update Makefile and Add test"

This reverts commit 330434069a4a6bd7d9b9b94c3e391317079af39a.

rniczh committed:
fix codecov

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
Update frontend/catalyst/compiler.py

Co-authored-by: Mehrdad Malek <39844030+mehrdad2m@users.noreply.github.com>

rniczh committed:
add llvm ir integeration test

rniczh committed:
ci

rniczh commented on frontend/catalyst/jit.py:
```diff
@@ -603,6 +603,10 @@ def __call__(self, *args, **kwargs):
 
+        # For llvm-ir target, compilation is complete, no execution needed
+        if self.compile_options.target == "llvmir":
+            return None
```
Thanks, I added the llvm ir test here: https://github.com/PennyLaneAI/catalyst/pull/2299/commits/e9e40923007496a9b70b7a97d0167906805c1eb6

rniczh commented on frontend/catalyst/jit.py:
```diff
@@ -820,6 +836,9 @@ def compile(self):
             shared_object, llvm_ir = self.compiler.run(self.mlir_module, self.workspace)
 
+        if self.compile_options.target == "llvmir":
+            return None, llvm_ir
```
Done https://github.com/PennyLaneAI/catalyst/pull/2299/commits/e9e40923007496a9b70b7a97d0167906805c1eb6

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
update test

mehrdad2m reviewed (approved):
🥳

dime10 commented on frontend/catalyst/jit.py:
```diff
@@ -603,6 +603,10 @@ def __call__(self, *args, **kwargs):
 
+        # For llvm-ir target, compilation is complete, no execution needed
+        if self.compile_options.target == "llvmir":
+            return None
```
This is not inline with how the existing `target` feature works, which could be confusing. `target` so far only limits AOT compilation, but here the function was invoked so compilation is done all the way in order to produce results. We should either change this behaviour globally, or remove these lines.

Do you have a need to call the qjit function if you don't want results? Note that `jit_compile` can be invoked directly if that's all you need.

dime10 commented on frontend/test/test_oqd/oqd/test_oqd_artiq_llvmir.py:
```diff
@@ -0,0 +1,94 @@
+            return qml.counts(all_outcomes=True)
+
+        # Get the LLVM IR
+        llvm_ir = circuit.qir
```
I think it's time we renamed this attribute `llvmir` 😅

rniczh replied:
Done, i changed all the `.qir` usage in the latest revision to `.llvmir`

dime10 commented on frontend/catalyst/jit.py:
```diff
@@ -656,6 +662,15 @@ def jit_compile(self, args, **kwargs):
         """
+        if self.compile_options.target == "llvmir":
+            if self.mlir_module is not None:
+                return False
```
Or is it because accessing the `.qir` attribute recompiles starting with the mlir_module anyway, in which case `target="mlir"` would be sufficient since you're not getting the llvmir from the first compilation anyway?

rniczh replied:
Beacuse I want to stop at the llvm ir when compiling with OQD, but as what you said, this behaviour not aligned with other `targets`, so I added a `link` option to qjit to determine whether we need to link to shared object. How do you think?
https://github.com/PennyLaneAI/catalyst/pull/2299/changes#diff-648249b9fca379a495b52fc5c000fb7c4425daa46b6044df3ddf4515b3bc7b0aR81

mehrdad2m commented on frontend/catalyst/jit.py:
> Is there a need here to split the compile stage either at the llvmir or object file generation, leaving the linking/executable generation to a separate stage? Would that prevent all these special branches?

I would love to see this happen. I didn't raise this because I thought maybe it's too much of a change in the core function for this PR, but I think that could help minimize the special branching. 

> Or can the feature get away with adding llmvir to the AOT pipeline only, without affecting the rest?

I think AOT should be enough for testing this, but is there a reason we don't want the `target` option in JIT mode? Personally, I would prefer to unify the behaviour.

dime10 replied:
Would you prefer that to raise an error then? Because if the compilation isn't forced then there is nothing to execute.
Or just return nothing? (I think this last option would be much more dangerous though)

rniczh committed:
ci

rniczh committed:
update llvmir repr logic

rniczh committed:
update llvmir logic

rniczh committed:
update test

rniczh committed:
update llvmir logic

rniczh committed:
update

rniczh committed:
update

rniczh commented on frontend/catalyst/jit.py:
```diff
@@ -621,16 +625,18 @@ def aot_compile(self):
 
-        if self.compile_options.target in ("binary",):
+        if self.compile_options.target in ("llvmir", "binary"):
             self.compiled_function, _ = self.compile()
```
I need the result of the compilation, the llvm ir, I cached the llvm ir before calling the `.llvmir`. https://github.com/PennyLaneAI/catalyst/pull/2299/changes#diff-648249b9fca379a495b52fc5c000fb7c4425daa46b6044df3ddf4515b3bc7b0aR637

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
remove redundant code

rniczh commented on frontend/catalyst/compiler.py:
```diff
@@ -484,16 +485,22 @@ def run_from_ir(self, ir: str, module_name: str, workspace: Directory):
+            output = output_ir_name if os.path.exists(output_ir_name) else None
+            if os.path.exists(tmp_infile_name):
+                os.remove(tmp_infile_name)
+            return output, out_IR
```
No, IR is clearly enough, it was reundnat. And we actually don't need to check whether the file exists. Fixed in https://github.com/PennyLaneAI/catalyst/pull/2299/commits/4c04388268bf7111c47776c1badb82af8c477bf7

```python
if not self.options.link:
            if os.path.exists(tmp_infile_name):
                os.remove(tmp_infile_name)
            return None, out_IR
```

rniczh commented on frontend/catalyst/compiler.py:
```diff
@@ -484,16 +485,22 @@ def run_from_ir(self, ir: str, module_name: str, workspace: Directory):
             out_IR = None
 
+        if not self.options.link:
+            output = output_ir_name if os.path.exists(output_ir_name) else None
```
Yes, but even it fails to lower to llvm ir or lower to llvm is not set, once the `link` option is not set, It's still cannot link whatsoever. And https://github.com/PennyLaneAI/catalyst/pull/2299#discussion_r2718805066 we just need to return out_ir for the case we don't want to link to object file.

dime10 replied:
Yeah the mechanism is safe I think, but how would you get it to actually produce the llvmir here?

rniczh committed:
remove unused file

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

dime10 reviewed (approved):
Little cleanup to avoid duplicate code

rniczh committed:
Update frontend/catalyst/compiler.py

Co-authored-by: David Ittah <dime10@users.noreply.github.com>

rniczh committed:
Update frontend/catalyst/compiler.py

Co-authored-by: David Ittah <dime10@users.noreply.github.com>

rniczh committed:
raise an error if no compiled function exists

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
Update check for OQD

rniczh committed:
test ci

rniczh committed:
fix

rniczh committed:
fix test

rniczh committed:
update

rniczh committed:
fix

rniczh committed:
formatting

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
ci

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
fix codecov

rniczh committed:
remove unused import

rniczh committed:
fix codecov

rniczh committed:
Add changelog

rniczh committed:
update

rniczh committed:
don't need to set llvmir explicitly for artiq

rniczh committed:
Set link=False if target=llvmir

rniczh committed:
update testcase

rniczh committed:
Change changelog's qjit decorator target to llvmir

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
test ci

rniczh committed:
update changelog

rniczh committed:
remove link entirely

rniczh committed:
add expected output

rniczh committed:
commit missing change for removal of link

rniczh committed:
clean redundant message

rniczh committed:
Add link

rniczh committed:
Add missing change

rniczh committed:
Add missing check

rniczh committed:
update

rniczh committed:
revert

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
Remove redundant

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
cache

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

dime10 commented on frontend/catalyst/jit.py:
```diff
@@ -838,6 +850,12 @@ def run(self, args, kwargs):
+            raise CompileError(
+                "Cannot execute function: no compiled function available. "
+                "The function must be compiled before execution."
+            )
```
Thanks, this is relatively clean given the circumstances 👌

dime10 reviewed (approved):
Alright I'll approve the qjit updates (note I haven't reviewed the OQD portions) because the changes are kept fairly small, however there is change in contract / behaviour of QJIT class in the following ways:
- `compile` is no longer guaranteed to output a `CompiledFunction` (wrapping binary code) which it was guaranteed to do until now
- `__call__` / `jit_compile` will no longer force compilation of the function down to binary if the target is set as `"llvmir"`, as a result `run` will raise an error for this case
  - this is different than the other target levels (`"jaxpr"`, `"mlir"`) which only control aot compilation, while calling the function will force the full compilation pipeline so the program can be executed
  - once the "llvmir" target is set there is no way to compile the program for execution, except by mutating the compile options in place or defining a new qjit object

The inconsistency could be solved by limiting the change (e.g. the internal link option) to take effect in AOT mode only. Alternatively, one could expand the target option to affect both AOT and and JIT mode for all targets, but I still think this comes with downsides (see the last bullet point above).
However, the change in the contract structure of QJIT (main idea: QJIT is broken down into discrete stages, each of which produces well defined results that can feed into the next stage) might only be resolvable by breaking the `compile` step up such that we have separate compile and link stages.

If @mehrdad2m is happy with the current state we can merge this PR but I think these two points will be added as technical debt.

rniczh committed:
revert the change that raises error for target=llvmir in jit mode

rniczh commented:
@mehrdad2m Sure, I reverted it in this commit: https://github.com/PennyLaneAI/catalyst/pull/2299/commits/67dba0996fdd989719a7c90b2e848facd0dd921d

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

mehrdad2m commented on frontend/catalyst/jit.py:
```diff
@@ -820,10 +828,12 @@ def compile(self):
         else:
             shared_object, llvm_ir = self.compiler.run(self.mlir_module, self.workspace)
 
+        if not self.compile_options.link:
```
Wouldn't this still crash the `jit_compile` path if we have `target=llvmir`? 🤔

rniczh replied:
Yeah, should remove those setting as well. The main goal is to remove all LLVM IR specializations in the pipeline. For now, we're reverting to only adding the link setting. Since end-to-end is not possible without disabling link, when using OQD's end-to-end, we need to change to actively construct using QJIT.

Changed: https://github.com/PennyLaneAI/catalyst/pull/2299/commits/9d80502ae78899f46a79a5c8d1fa031bccd50258

Therefore, in terms of usage, we're changing the approach to first create a qnode with `@qnode`, then use `QJIT` with the current compile options to achieve this.

```python
oqd_pipelines = _get_oqd_pipelines()

@qml.set_shots(4)
@qml.qnode(oqd_dev)
def circuit():
    x = np.pi / 2
    qml.RX(x, wires=0)
    return qml.counts(all_outcomes=True)

# Get the LLVM IR
compiled_circuit = QJIT(circuit, CompileOptions(link=False, pipelines=oqd_pipelines))
llvm_ir = compiled_circuit.llvmir
```

mehrdad2m replied:
Thanks @rniczh . I think this looks good for now. This could be improved if we divide `compile()` into `compile()` and `link()` like @dime10 suggested. But we don't need to address this in this PR.

rniczh committed:
no specialize on target=llvmir

rniczh committed:
update test

rniczh committed:
format

rniczh committed:
remove unused import

rniczh committed:
formatting

rniczh committed:
formatting

rniczh commented:
> Did we remove a test for `compile_to_artiq`? Codecov is not happy about it, and it seems to be not used anywhere except in documentation.

We had the `ARM` target and `lld` installed for OQD pipeline at the beginning (in `mlir/CMakeLists.txt`, but after discussing with @paul0403, we decided to remove them (less intrusive). So that means we can't have the test for the `compile_to_artiq` function.

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
update pass for latest llvm update

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline

rniczh committed:
remove  function (moved to third-party repo instead)

rniczh committed:
update changelog

rniczh commented:
> > > Did we remove a test for `compile_to_artiq`? Codecov is not happy about it, and it seems to be not used anywhere except in documentation.
> > 
> > 
> > We had the `ARM` target and `lld` installed for OQD pipeline at the beginning (in `mlir/CMakeLists.txt`, but after discussing with @paul0403, we decided to remove them (less intrusive). So that means we can't have the test for the `compile_to_artiq` function.
> > EDIT: so this happend in this commit: [354ad5c](https://github.com/PennyLaneAI/catalyst/commit/354ad5ce2f4089522a15c0409650532895df2429)
> 
> Ah I see, thanks for the context. This makes me wonder if this file should live somewhere else like in the [oqd-catalyst-integration-test](https://github.com/OpenQuantumDesign/oqd-catalyst-integration-test/tree/main) repo. Especially that it is not relevant for any general user. Having a completely untested file like this make me a little uncomfortable. What do you think? We can maybe discuss this tomorrow in our standup.

As we discussed in offline. We removed the `compile_to_artiq` function from this PR and moved to the third-party OQD repo. So this PR only reponsible for generating llvm ir.

rniczh committed:
for jit, it's alawys true

mehrdad2m reviewed (approved):
Nice! Thanks for the great work! 🚀

rniczh committed:
Merge branch 'main' into rniczh/add-new-oqd-pipeline
